### PR TITLE
[TypeChecker] `<unknown>` diagnostic location regarding `Codable` derived conformances

### PR DESCRIPTION
**Pull request by HassanElDesouky**
Fixing `<unknown>` diagnostic location regarding `Codable` derived conformances.

cc @LucianoPAlmeida, @CodaFi, @xedin  :) 

rdar://problem/59655704
Resolves [SR-12248](https://bugs.swift.org/browse/SR-12248).

### CODE REVIEW

File: test/decl/protocol/special/coding/struct_codable_simple.swift [hunk 1]
```diff
@@ -29,3 +29,21 @@ let _ = SimpleStruct.encode(to:)
 // The synthesized CodingKeys type should not be accessible from outside the
 // struct.
 let _ = SimpleStruct.CodingKeys.self // expected-error {{'CodingKeys' is inaccessible due to 'private' protection level}}
+
+// rdar://problem/59655704 
+struct SR_12248_1: Codable { // expected-error {{type 'SR_12248_1' does not conform to protocol 'Encodable'}}
+  var x: Int // expected-note {{'x' previously declared here}}
+  var x: Int // expected-error {{invalid redeclaration of 'x'}}
+  // expected-note@-1 {{cannot automatically synthesize 'Encodable' because '<<error type>>' does not conform to 'Encodable'}}
```
Comment (LucianoPAlmeida) [inline test/decl/protocol/special/coding/struct_codable_simple.swift:37]: Not sure if this is a requirement for the SR, but I think this shouldn't be noted when the synthesize type `hasError()`, in fact, the missing `Encodable` conformance diagnostic itself shouldn't be diagnosed in this case.
WDYT?

Comment (HassanElDesouky) [reply test/decl/protocol/special/coding/struct_codable_simple.swift:37]: Honestly, I also thought of the same thing. I don’t think this needs to be tested. However, as a newbie I decided to add a lot of tests 😅

Comment (LucianoPAlmeida) [reply test/decl/protocol/special/coding/struct_codable_simple.swift:37]: You could take a look at `DerivedConformance::deriveEncodable`, note that in 

https://github.com/apple/swift/blob/426583523e063be1ed75027044f56873b7b73183/lib/Sema/TypeCheckDeclPrimary.cpp#L721 

The member is flagged as invalid, so on `DerivedConformance::deriveEncodable` and decodable you could check for any nominal member and skip the conformance diagnostic if any of the members are invalid :)

Comment (CodaFi) [reply test/decl/protocol/special/coding/struct_codable_simple.swift:37]: The `ErrorType`s are unfortunate, but it's not enough to block the PR.

**Comment by CodaFi** [comment]
@swift-ci test

**Review by CodaFi — APPROVED** [review]

**Comment by CodaFi** [comment]
@swift-ci test Windows

**Comment by CodaFi** [comment]
@HassanElDesouky Could you take a look at removing all of the `setInvalid()s` from redeclaration checking? It might just fix the `ErrorType` thing, but it would be a good cleanup to make here because it's really just resetting the interface type and not actually setting a bit anymore.

**Comment by HassanElDesouky** [comment]
> @HassanElDesouky Could you take a look at removing all of the `setInvalid()s` from redeclaration checking? It might just fix the `ErrorType` thing, but it would be a good cleanup to make here because it's really just resetting the interface type and not actually setting a bit anymore.

@CodaFi Just so I understand... You mean remove all of `setInvalid()s` **ONLY** in this function?
https://github.com/apple/swift/blob/426583523e063be1ed75027044f56873b7b73183/lib/Sema/TypeCheckDeclPrimary.cpp#L483

Or do you mean remove all of `setInvalid()s` in the file?

**Comment by CodaFi** [comment]
Just in this function would be a good start.  Generally, requests that reset the interface type are also candidates for having their `setInvalid` calls removed, but we’ll start small eh?

**Comment by HassanElDesouky** [comment]
Yeah, it's better if we start small. I'll go ahead and remove the `setInvalid`.

**Comment by CodaFi** [comment]
Just to be clear, I'm going to merge this. Please shoot us another PR.

**Comment by HassanElDesouky** [comment]
Thanks for clearing this out. But since it will be a new PR, maybe remove all of `setInvalid()s` in the TypeCheckDeclPrimary file?

Just to be clear about my task... 
1. I will remove all of `setInvalid()s` and won't substitute them with anything.
2. Since it's only a cleanup, I won't need to update/add new tests.
Is that right?

**Comment by CodaFi** [comment]
> I will remove all of setInvalid()s and won't substitute them with anything.

Yep. Should be a nice sea of red.

> Since it's only a cleanup, I won't need to update/add new tests.
Is that right?

You might need to go update some tests if the diagnostics change, but I wouldn't anticipate much beyond that.

**Comment by HassanElDesouky** [comment]
Great! I'll make sure to do that once I wake up *later this day* as it's 4:37 am here in Cairo :) 
Thank you <3

**Comment by HassanElDesouky** [comment]
@CodaFi  Is the failure related to my modification?

**Comment by CodaFi** [comment]
Yeah. It's just a few tests that no longer emit notes

```
swift/test/Sema/enum_conformance_synthesis.swift:64:42: error: expected note not produced
   var hashValue: String { return "" } // expected-note {{previously declared here}}
```

```
19:56:51   // expected-note@-2 {{'$i' previously declared here}}
19:56:51 ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
19:56:51 /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/swift/test/decl/var/property_wrappers.swift:841:6: error: expected note not produced
19:56:51   // expected-note@-3 {{'_i' previously declared here}}
19:56:51 ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
19:56:51 /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/swift/test/decl/var/property_wrappers.swift:844:6: error: expected error not produced
19:56:51   // expected-error@-2 {{invalid redeclaration of '$i'}}
19:56:51 ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
19:56:51 /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/swift/test/decl/var/property_wrappers.swift:845:6: error: expected error not produced
19:56:51   // expected-error@-3 {{invalid redeclaration of '_i'}}
```

Both of those are good to remove.

**Comment by HassanElDesouky** [comment]
@CodaFi I'm sorry I don't know how to edit this correctly so, I'll close this PR and open a new one :)

**Pull request closed without merging**